### PR TITLE
Fixes #2640 error when clearing cache on O2Switch when using PHP7.1+

### LINKED ISSUE #2640: Error when clearing cache on O2Switch when using PHP7.1+
**Describe the bug**
We're getting the following error when clearing the post cache on O2Switch with PHP7.1+

`Uncaught Error: [] operator not supported for strings in /home/domain/public_html/wp-content/plugins/wp-rocket-compat-varnish-ip/wp-rocket-compat-varnish-ip.php:29`

That's because the short array push syntax is not allowed on a string or non-existing variables since PHP7.1.

I was able to fix it in the helper plugin by declaring the variable as an array:
https://github.com/wp-media/wp-rocket-helpers/blob/9f7186010f3d405d0b614cafd008ba359006f1ec/compatibility/wp-rocket-compat-varnish-ip/wp-rocket-compat-varnish-ip.php#L30

However, it ended up with the following notice each time the post is cleaned (14+ times on post update):
`PHP Notice:  Array to string conversion in /home/domain/public_html/wp-content/plugins/wp-rocket/inc/3rd-party/hosting/o2switch.php on line 122`

This problem happened after updating our helper to be compatible with multiple Varnish IPs.

**To Reproduce**
1. On a site on O2Switch enable their caching
2. Make sure that PHP version is at least 7.1
3. Enable Varnish Add-on
4. Install and configure the helper:
https://fr.docs.wp-rocket.me/article/1113-utiliser-wp-rocket-avec-xtreme-cache-o2switch
5. Clear the post cache and check PHP logs

**Expected behavior**
Correctly cleaned cache without errors nor notices

**Additional context**
We could create a specific Varnish helper plugin for O2Switch, since they're using only one IP, the array is not needed.

The other solution is to fix this problem on the WP Rocket's end.

Ticket:
https://secure.helpscout.net/conversation/1158506945/163617/

**Backlog Grooming (for WP Media dev team use only)**
- [x] Reproduce the problem
- [x] Identify the root cause
- [x] Scope a solution
- [x] Estimate the effort

### PR DESCRIPTION
**Pull request by engahmeds3ed**
Fixes #2640 

- [x] Migrate the o2switch file into our new architecture
- [x] Remove duplicated code with varnish class
- [x] Unit Tests
- [x] Integration Tests

### Acceptance criteria
- Varnish add-on is correctly showing that we are on o2switch and auto-enabled
- Varnish cache is correctly cleared when clearing WP Rocket cache, without errors in the log and on the site

### CODE REVIEW

**Comment by engahmeds3ed** [comment]
@wp-media/qa @Davidacu this PR is ready for QA, thanks

**Comment by Tabrisrp** [comment]
This PR will need to be updated after PRs #2860 & #2874

File: inc/ThirdParty/Hostings/O2Switch.php [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+	 * @param array $headers Headers to send.
+	 *
+	 * @return array Array for headers to be sent.
+	 */
+	public function add_purge_headers( $headers ) {
+		$headers['X-VC-Purge-Key'] = O2SWITCH_VARNISH_PURGE_KEY;
+
+		if ( 'regex' === $headers['X-Purge-Method'] ) {
+			$headers['X-Purge-Regex'] = '.*';
+		}
```
Comment (Tabrisrp) [inline inc/ThirdParty/Hostings/O2Switch.php:84]: You can also unset `$headers['X-Purge-Method']` at the same time to make sure we don't send unnecessary headers

Comment (engahmeds3ed) [reply inc/ThirdParty/Hostings/O2Switch.php:84]: Ok, sure

**Comment by engahmeds3ed** [comment]
@Tabrisrp Done, please take a look before calling QA.
- Solve branch conflicts
- Adjust the code to match the new activation/deactivation design
- Introduce the new filter for purge_url.

File: inc/Addon/Varnish/Varnish.php [hunk 2]
```diff
@@ -111,8 +111,23 @@ public function purge( $url ) {
...
-			$host      = ! empty( $ip ) ? $ip : str_replace( '*', '', $parse_url['host'] );
-			$purge_url = $scheme . '://' . $host . $parse_url['path'] . $regex;
+			$host           = ! empty( $ip ) ? $ip : str_replace( '*', '', $parse_url['host'] );
+			$purge_url_main = $scheme . '://' . $host . $parse_url['path'];
+
+			/**
+			 * Filters the purge url.
+			 *
+			 * @since 3.6.3
+			 *
```
Comment (Tabrisrp) [inline inc/Addon/Varnish/Varnish.php:121]: One parameter is missing for the first one containing the full URL

Comment (engahmeds3ed) [reply inc/Addon/Varnish/Varnish.php:121]: Really you are awesome.
Done

**Comment by engahmeds3ed** [comment]
@Tabrisrp Plz final review before sending it to QA, thanks.

**Comment by Tabrisrp** [comment]
@wp-media/qa Ready for QA, acceptance criteria in the OP

**Pull request merged**